src/mds/CDir.cc

```diff
@@ -2187,7 +2185,7 @@ void CDir::_encode_dentry(CDentry *dn, bufferlist& bl,
     bl.append('L');         // remote link
     ::encode(ino, bl);
     ::encode(d_type, bl);
-  } else {
+  } else if (dn->linkage.is_primary()) {
```

Comment:
Same as above: if we really never expect a null dentry here, add an assertion before or after the if block